### PR TITLE
JIT: Stop allocating address exposed locals if they are unreferenced

### PR DESCRIPTION
**Pull request by jakobbotsch**
I don't know why this would be necessary, so if it really is then hopefully tests will let me know.

### CODE REVIEW

**Comment by jakobbotsch** [comment]
The failure is #87837.

cc @dotnet/jit-contrib PTAL @BruceForstall.

[Diffs](https://dev.azure.com/dnceng-public/public/_build/results?buildId=312951&view=ms.vss-build-web.run-extensions-tab)

Any concerns? No references means that nothing can have taken the address of the local, so the only way to access it would be indirectly by making assumptions about the stack layout and neighboring locals.

File: src/coreclr/jit/regalloc.cpp [hunk 1]
```diff
@@ -342,18 +333,7 @@ void Compiler::raMarkStkVars()
...
-                    varDsc->lvImplicitlyReferenced = 1;
-                }
-
-                needSlot |= true;
-
-                if (!varDsc->lvIsParam)
-                {
-                    varDsc->lvMustInit = true;
-                }
+                assert(!"unreferenced local in debug codegen");
```
Comment (BruceForstall) [inline src/coreclr/jit/regalloc.cpp:336]: `needSlot` is always `false` now; you could remove it entirely.

**Review by BruceForstall — COMMENTED** [review]
Seems good. I know there are a few implicit ordering dependencies of frame data on x86 but IIRC, not things that have local vars representing them (or maybe we force the ordering of those special local vars).

File: src/coreclr/jit/regalloc.cpp (hunk 1 above)
Comment (jakobbotsch) [reply src/coreclr/jit/regalloc.cpp:336]: Ah, good catch

Comment (jakobbotsch) [reply src/coreclr/jit/regalloc.cpp:326]: I modernized the styling of the whole file and removed the outdated comment at the top as well.

Comment (jakobbotsch) [reply src/coreclr/jit/regalloc.cpp]: I also removed the assert here as it doesn't seem that valuable, but let me know if you disagree.

**Review by BruceForstall — APPROVED** [review]
LGTM

**Pull request merged**